git/cmd.py

```diff
@@ -603,6 +610,73 @@ def execute(self, command,
         else:
             return stdout_value
 
+    def set_environment(self, **kwargs):
+        """
+        Set environment variables for future git invocations. Return all changed
+        values in a format that can be passed back into this function to revert
+        the changes:
+
+        ``Examples``::
+
+            old_env = self.set_environment(PWD='/tmp')
+            self.set_environment(**old_env)
+
+        :param kwargs: environment variables to use for git processes
+        :return: dict that maps environment variables to their old values
+        """
+        old_env = {}
+        for key, value in kwargs.iteritems():
+            # set value if it is None
+            if value is not None:
+                if key in self._environment:
+                    old_env[key] = self._environment[key]
+                else:
+                    old_env[key] = None
+                self._environment[key] = value
+            # remove key from environment if its value is None
+            elif key in self._environment:
+                old_env[key] = self._environment[key]
+                del self._environment[key]
+        return old_env
+
+    @contextmanager
+    def environment(self, **kwargs):
+        """
+        A context manager around the above set_environment to restore the
+        environment back to its previous state after operation.
+
+        ``Examples``::
+
+            with self.environment(GIT_SSH='/bin/ssh_wrapper'):
+                repo.remotes.origin.fetch()
+
+        :param kwargs: see set_environment
+        """
+        old_env = self.set_environment(**kwargs)
+        try:
+            yield
+        finally:
+            self.set_environment(**old_env)
+
+    @contextmanager
+    def sshkey(self, sshkey_file):
```

Comment:
For a moment I thought this method is to specific, but on the other hand, it makes setting ssh-keys (by using key-files) really easy.
Therefore the only thing I will do is see how this works on windows, and possibly put in an assertion to run on non-windows systems only.